compaction/compaction.cc

```diff
@@ -772,6 +773,30 @@ class compaction {
         };
     }
 
+    std::function<api::timestamp_type(const dht::decorated_key&)> max_repair_purgeable_func() {
+        if (!tombstone_expiration_enabled()) {
+            return [] (const dht::decorated_key& dk) {
+                return api::min_timestamp;
+            };
+        }
+        return [this] (const dht::decorated_key& dk) {
+            auto ts = _cf.get_repair_time(dk.token());
+            if (ts == api::missing_timestamp) {
+                clogger.info("Get repair time for ks={}, table={}, dk={}, repair_timestamp={}, ret={}, missing",
+                        _cf.schema()->ks_name(), _cf.schema()->cf_name(), dk, ts, api::min_timestamp);
+                return api::min_timestamp;
+            } else {
+                // TODO: Make propagation_delay configurable
+                constexpr api::timestamp_type propagation_delay =
+                        std::chrono::duration_cast<api::timestamp_clock::duration>(std::chrono::seconds(10)).count();
+                auto ret = ts > propagation_delay ? ts - propagation_delay : api::min_timestamp;
```

Comment:
I don't understand how we can compare `ts` (a timestamp) with `propagation_delay` which is a duration really? I also don't understand why we are returning `api::min_timestamp` when the propagation delay hasn't passed yet. Shouldn't propagation delay define a window in which we still keep tombstones (because we can get writes shadowed by them), but tombstones much older than than can be expired? This is how gc grace period works too. So we should return something like:
```
return ts - propagation_delay;
```

Reply:
I should have used saturating_subtract.  We can not simply use ts - propagaton_deay, because the ts can be very small, so ts - propagaton_deay will be smaller than  api::min_timestamp, which makes no sense.